drivers/media/i2c/irs1125.c

```diff
@@ -0,0 +1,1137 @@
+{
+    struct v4l2_ctrl *ctrl;
+    int err, i;
+    struct v4l2_ctrl_handler *hdl;
+    
+    hdl = &sensor->ctrl_handler;
+    v4l2_ctrl_handler_init(hdl, ARRAY_SIZE(irs1125_custom_ctrls));
+
+    for (i = 0; i < ARRAY_SIZE(irs1125_custom_ctrls); i++)
+    {
+        ctrl = v4l2_ctrl_new_custom(hdl, &irs1125_custom_ctrls[i], NULL);
+        if (ctrl == NULL)
+        {
+            dev_err(dev, "Failed to init custom control %s\n", irs1125_custom_ctrls[i].name);
+        }
+        else if (irs1125_custom_ctrls[i].id == IRS1125_CID_NUM_SEQS)
+        {
+                sensor->ctrl_numseq = ctrl;
+        }
+        else if (irs1125_custom_ctrls[i].id == IRS1125_CID_MOD_PLL)
+        {
+                sensor->ctrl_modplls = ctrl;
+        }
+    }
+    
+    
+    if (hdl->error)
+    {
+        dev_err(dev, "Error %d adding controls\n",
+            hdl->error);
+        err = hdl->error;
+        goto error_ctrls;
+    }
+    
+    sensor->sd.ctrl_handler = hdl;
+    return 0;
+    
+error_ctrls:
+    v4l2_ctrl_handler_free(&sensor->ctrl_handler);
+    return -err;
+}
+
+static int irs1125_ident_setup(struct irs1125 *sensor, struct device *dev)
+{
+    int ret;
+    struct v4l2_ctrl *ctrl;
+    struct v4l2_subdev *sd;
+    u16 read;
+
+
+    sd = &sensor->sd;
+
+    ctrl = v4l2_ctrl_find(&sensor->ctrl_handler, IRS1125_CID_IDENT0);
+    if (!ctrl) {
+        dev_err(dev,
+            "could not find device ctrl.\n");
+        return -EINVAL;
+    }
+
+    ret = irs1125_read(sd, IRS1125_REG_EFUSEVAL2, &read);
+    if (ret < 0)
+    {
+        dev_err(dev, "error reading from i2c\n");
+        return -EIO;
+    }
+
+    v4l2_ctrl_s_ctrl(ctrl, read);
+
+    ctrl = v4l2_ctrl_find(&sensor->ctrl_handler, IRS1125_CID_IDENT1);
+    if (!ctrl) {
+        dev_err(dev,
+            "could not find device ctrl.\n");
+        return -EINVAL;
+    }
+
+    ret = irs1125_read(sd, IRS1125_REG_EFUSEVAL3, &read);
+    if (ret < 0)
+    {
+        dev_err(dev, "error reading from i2c\n");
+        return -EIO;
+    }
+
+    v4l2_ctrl_s_ctrl(ctrl, read);
+
+    ctrl = v4l2_ctrl_find(&sensor->ctrl_handler, IRS1125_CID_IDENT2);
+    if (!ctrl) {
+        dev_err(dev,
+            "could not find device ctrl.\n");
+        return -EINVAL;
+    }
+
+    ret = irs1125_read(sd, IRS1125_REG_EFUSEVAL4, &read);
+    if (ret < 0)
+    {
+        dev_err(dev, "error reading from i2c\n");
+        return -EIO;
+    }
+    v4l2_ctrl_s_ctrl(ctrl, read&0xFFFC);
+    
+    return 0;
+}
+
+static int irs1125_probe(struct i2c_client *client,
+            const struct i2c_device_id *id)
+{
+    struct device *dev = &client->dev;
+    struct irs1125 *sensor;
+    int ret;
+    struct v4l2_subdev *sd;
+    struct device_node *np = client->dev.of_node;
+    u32 xclk_freq;
+    int gpio_num;
+
+    sensor = devm_kzalloc(dev, sizeof(*sensor), GFP_KERNEL);
+    if (!sensor)
+        return -ENOMEM;
+
+    if (IS_ENABLED(CONFIG_OF) && np) {
```

Comment:
Do you support configuration without device tree? I suspect not, therefore checking CONFIG_OF is selected is redundant.

I think ov5647 has been a bad example in this case. I merged an imx219 driver about a month ago which is hopefully a better example.

Reply:
Okay, I followed the example of imx219. Should be better now